src/rebalancers/uniswap-v3/UniswapV3Rebalancer.sol

```diff
@@ -0,0 +1,447 @@
+    mapping(address owner => address initiator) public ownerToInitiator;
+
+    // A mapping from initiator to rebalancing fee.
+    mapping(address initiator => InitiatorInfo) public initiatorInfo;
+
+    // A struct with the state of a specific position, only used in memory.
+    struct PositionState {
+        address pool;
+        address token0;
+        address token1;
+        uint24 fee;
+        int24 newUpperTick;
+        int24 newLowerTick;
+        uint128 liquidity;
+        uint256 sqrtPriceX96;
+        uint256 sqrtRatioLower;
+        uint256 sqrtRatioUpper;
+        uint256 lowerBoundSqrtPriceX96;
+        uint256 upperBoundSqrtPriceX96;
+    }
+
+    struct InitiatorInfo {
+        uint256 upperSqrtPriceDeviation;
+        uint256 lowerSqrtPriceDeviation;
+        uint256 fee;
+    }
+
+    /* //////////////////////////////////////////////////////////////
+                                ERRORS
+    ////////////////////////////////////////////////////////////// */
+
+    error DecreaseFeeOnly();
+    error DecreaseToleranceOnly();
+    error FeeAlreadySet();
+    error LiquidityTresholdExceeded();
+    error MaxInitiatorFee();
+    error MaxTolerance();
+    error NotAnAccount();
+    error OnlyAccount();
+    error OnlyPool();
+    error Reentered();
+    error UnbalancedPool();
+    error InitiatorNotValid();
+
+    /* //////////////////////////////////////////////////////////////
+                                EVENTS
+    ////////////////////////////////////////////////////////////// */
+
+    event Rebalance(address indexed account, uint256 id);
+
+    /* //////////////////////////////////////////////////////////////
+                            CONSTRUCTOR
+    ////////////////////////////////////////////////////////////// */
+
+    /**
+     * @param liquidityTreshold .
+     */
+    constructor(uint256 liquidityTreshold) {
+        // TODO: max treshold ?
+        LIQUIDITY_TRESHOLD = liquidityTreshold;
+    }
+
+    /* ///////////////////////////////////////////////////////////////
+                             REBALANCING LOGIC
+    /////////////////////////////////////////////////////////////// */
+
+    /**
+     * @notice
+     * @param account_ The Arcadia Account owning the position.
+     * @param id The id of the Liquidity Position.
+     * @param lowerTick .
+     * @param upperTick .
+     * @dev When both lowerTick and upperTick are zero, ticks will be updated with same tick-spacing as current position.
+     */
+    function rebalancePosition(address account_, uint256 id, int24 lowerTick, int24 upperTick) external {
+        // Store Account address, used to validate the caller of the executeAction() callback.
+        if (account != address(0)) revert Reentered();
+        if (!ArcadiaLogic.FACTORY.isAccount(account_)) revert NotAnAccount();
+        if (ownerToInitiator[IAccount(account_).owner()] != msg.sender) revert InitiatorNotValid();
```

Comment:
This would require a double tx by the user? Once to set the owner to initiator, once for the flash action?
This ownerToInitiator would then also be non-account specific?

Reply:
we could do it account-specific, but then when transferring an Account, they are transferred with the AM's attached to it. Don't think it's optimal but we can discuss which is best

Reply:
or I can add a nested mapping owner => account => initiator ?

Reply:
> This would require a double tx by the user? Once to set the owner to initiator, once for the flash action? This ownerToInitiator would then also be non-account specific?

No user only has to set once an initiator per account. Flash action trigerred by initiator.